ecosystem/node-checker/fn-check-client/src/check.rs

```diff
@@ -0,0 +1,396 @@
+        Self {
+            result,
+            timestamp: now,
+            node_url,
+        }
+    }
+}
+
+/// We use this struct to capture the result of checking a node, or the lack thereof.
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum SingleCheckResult {
+    /// The node was successfully checked. Note: The evaulation itself could
+    /// still indicat, a problem with the node, this just states that we were
+    /// able to check the node sucessfully with NHC.
+    Success(SingleCheckSuccess),
+
+    /// Something went wrong with checking the node with NHC.
+    NodeCheckFailure(NodeCheckFailure),
+
+    /// The account does not have a VFN registered on chain.
+    NoVfnRegistered(NoVfnRegistered),
+
+    /// The network address could not be deserialized.
+    CouldNotDeserializeNetworkAddress(CouldNotDeserializeNetworkAddress),
+
+    /// The network address was incomplete, e.g. missing an IP, port, public
+    /// key, etc.
+    IncompleteNetworkAddress(IncompleteNetworkAddress),
+}
+
+#[derive(Debug, Serialize)]
+pub struct SingleCheckSuccess {
+    /// The evaluation summary returned by NHC. This doesn't necessarily imply
+    /// that the node passed the evaluation, just that an evaluation was returned
+    /// successfully and it passed the API available check.
+    pub evaluation_summary: EvaluationSummary,
+
+    /// This is the address that we used to get this successful evaluation.
+    /// This is presented in a normal URL format, not the NetworkAddress
+    /// representation. Example value for this field: http://65.109.17.29:8080.
+    /// Note, sometimes the address we started with was a DNS name, and we resolved
+    /// it to an IP address. As such, this IP address may become incorrect down
+    /// the line. In that case, refer to fn_address in SingleCheck, or just
+    /// run this tool again.
+    pub fn_address_url: String,
+}
+
+impl SingleCheckSuccess {
+    pub fn new(evaluation_summary: EvaluationSummary, fn_address_url: String) -> Self {
+        Self {
+            evaluation_summary,
+            fn_address_url,
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct NodeCheckFailure {
+    pub message: String,
+    pub code: NodeCheckFailureCode,
+}
+
+impl NodeCheckFailure {
+    pub fn new(message: String, code: NodeCheckFailureCode) -> Self {
+        Self { message, code }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub enum NodeCheckFailureCode {
+    // Something went wrong when sending / receiving the request.
+    RequestResponseError,
+
+    // The response from NHC was not a 200, implying a problem with NHC.
+    ResponseNot200,
+
+    // The response from NHC couldn't be deserialized.
+    CouldNotDeserializeResponse,
+
+    // NHC returned an evaluation that indicates that the API port is closed.
+    ApiPortClosed,
+}
+
+// These are necessary because we can't just use a unit type for this enum variant.
```

Comment:
Silly question: why not? 😄 Is it because each variant has different properties or something else? If it's the former, something like this should work:

```
pub enum Example {
    NoVfnRegistered,
    CouldNotDeserializeNetworkAddress(String),
    IncompleteNetworkAddress(String)
}
```

Reply:
It's because I'm serializing the enum using internal tagging, in which case serde requires that each field have something within them.

Reply:
I'll update the comment.